### DISCUSSION TITLE
Foreign keys in directus_revisions table

### DISCUSSION THREAD

**Discussion by Cyon85** [directus/directus#13481] [Feature Requests]
There are two foreign keys in the directus_revisions table at the moment:
- directus_revisions_activity_foreign
- directus_revisions_parent_foreign

for the key "directus_revisions_activity_foreign" it says "ON DELETE CASCADE"
that works fine if you have a revision where the "parent" column is "null"

if the parent column is not null, for example there is an activity where you created a file in your collection,
it ends up with something like this when you want to delete the activity:

`ERROR: update or delete on table "directus_revisions" violates foreign key constraint "directus_revisions_parent_foreign" on table "directus_revisions" Detail: Key (id)=(597702) is still referenced from table "directus_revisions".
`
to prevent that error, isnt it better to set the parent column to null ?

```sql
alter table directus_revisions
    add constraint directus_revisions_parent_foreign
        foreign key (parent) references directus_revisions
            on delete set null;
```

**Comment by rijkvanzanten** [comment]
I agree, but that doesn't work reliable in every SQL vendor. MS SQL for example prevents creating foreign key constraint triggers recursively to the same table, and vendors that do support it aren't always reliable

**Reply by Cyon85** [reply]
Because it is not reliable, we won't it use for the SQL vendor it works?

Puh, I know the discussions related to the activity and revisions table ... 
Without a custom schedule-hook there is no chance to implement some retention rules for activities and there revisions.
And the easiest way to implement retention rules for a collection is, delete an activity that is older then ...
That is only possible if there is no relation to that collection, without some extra code to remove the foreign key parent first.

At the moment i did a custom-migration:


```
module.exports = {
	async up(knex) {
		await knex.schema.alterTable('directus_revisions', (table) => {
			table.dropForeign(['parent']);
		});

		await knex.schema.alterTable('directus_revisions', (table) => {
			table.foreign('parent').references('directus_revisions.id').onDelete('SET NULL');
		});
	},

	async down(knex) {
		await knex.schema.alterTable('directus_revisions', (table) => {
			table.dropForeign(['parent']);
		});

		await knex.schema.alterTable('directus_revisions', (table) => {
			table.foreign('parent').references('directus_revisions.id');
		});
	},

};
```


after that, something like this is possible:


```
			await activityService.deleteMany(
				await getReadyToDeleteActivities(action, collection, pastTime, limit)
					.then((res) => {
						return res.map(function (elem) {
							return elem.id;
						});
					})
			);
```

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
